agrona/src/main/java/org/agrona/concurrent/UnsafeBuffer.java

```diff
@@ -515,6 +557,49 @@ public int addIntOrdered(final int index, final int increment)
         return UnsafeApi.getAndAddIntRelease(byteArray, addressOffset + index, increment);
     }
 
+
+    /**
+     * {@inheritDoc}
+     */
+    public void putIntOpaque(final int index, final int value)
+    {
+        if (SHOULD_BOUNDS_CHECK)
+        {
+            boundsCheck0(index, SIZE_OF_LONG);
+        }
+
+        UnsafeApi.putIntOpaque(byteArray, addressOffset + index, value);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public int getIntOpaque(final int index)
+    {
+        if (SHOULD_BOUNDS_CHECK)
+        {
+            boundsCheck0(index, SIZE_OF_LONG);
+        }
+
+        return UnsafeApi.getIntOpaque(byteArray, addressOffset + index);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public int addIntOpaque(final int index, final int increment)
+    {
+        if (SHOULD_BOUNDS_CHECK)
+        {
+            boundsCheck0(index, SIZE_OF_LONG);
+        }
+
+        final int oldValue = UnsafeApi.getIntOpaque(byteArray, addressOffset + index);
+        final int newValue = oldValue + increment;
+        UnsafeApi.putIntOpaque(byteArray, addressOffset + index, newValue);
+        return oldValue;
+    }
+
```

Comment:
To be correct this method needs to use a CAS loop similarly to how `addIntOrdered` is implemented. The latter is invoking `jdk.internal.misc.Unsafe#getAndAddIntRelease` which in Java code looks like this:
```java
    @ForceInline
    public final int getAndAddIntRelease(Object o, long offset, int delta) {
        int v;
        do {
            v = getInt(o, offset);
        } while (!weakCompareAndSetIntRelease(o, offset, v, v + delta));
        return v;
    }
```
Funny fact, this method is not an intrinsic unlike `getAndAddInt` method.

Reply:
It doesn't make sense for the addIntOpaque to be atomic, while addIntOrdered isn't. Opaque is weaker than ordered.

```     
public int addIntOrdered(final int index, final int increment)
    {
        if (SHOULD_BOUNDS_CHECK)
        {
            boundsCheck0(index, SIZE_OF_INT);
        }

        final long offset = addressOffset + index;
        final byte[] array = this.byteArray;
        final int value = UNSAFE.getInt(array, offset);
        UNSAFE.putOrderedInt(array, offset, value + increment);

        return value;
    }
```

Reply:
Fair point. I got confused as I assumed all methods in `AtomicBuffer` should be atomic.